runtime/compiler/optimizer/EscapeAnalysis.cpp

```diff
@@ -3742,32 +3742,60 @@ static bool isFinalizableInlineTest(TR::Compilation *comp, TR::Node *candidate,
    TR_J9VMBase *fej9 = (TR_J9VMBase *)(comp->fe());
 
    bool is64Bit = TR::Compiler->target.is64Bit();
-   TR::ILOpCodes ifOp = is64Bit ? TR::iflcmpne : TR::ificmpne;
-   TR::ILOpCodes andOp = is64Bit ? TR::land : TR::iand;
+
+   // root
+   //   r1      - first child of root
+   //     r11   - first child of first child of root
+   //     r12   - second child of first child of root
+   //   r2      - second child of root
+   TR::Node *r1 = (root->getNumChildren() > 0) ? root->getFirstChild() : NULL;
+   TR::Node *r2 = (root->getNumChildren() > 1) ? root->getSecondChild() : NULL;
+   TR::Node *r11 = (r1 && r1->getNumChildren() > 0) ? r1->getFirstChild() : NULL;
+   TR::Node *r12 = (r1 && r1->getNumChildren() > 1) ? r1->getSecondChild() : NULL;
+
+   bool castDownToInt = is64Bit && r11 && (r11->getOpCodeValue() == TR::l2i);
+   bool usesLongOps = is64Bit && !castDownToInt;
+
+   TR::ILOpCodes ifOp = usesLongOps ? TR::iflcmpne : TR::ificmpne;
+   TR::ILOpCodes andOp = usesLongOps ? TR::land : TR::iand;
    TR::ILOpCodes loadOp = is64Bit ? TR::lloadi : TR::iloadi;
-   TR::ILOpCodes constOp = is64Bit ? TR::lconst : TR::iconst;
+   TR::ILOpCodes constOp = usesLongOps ? TR::lconst : TR::iconst;
+
+   TR::Node *loadNode = castDownToInt ? r11->getFirstChild() : r11;
 
    /*
       Looking for a pattern of:
```

Comment:
FYI if you did not know we have quite a handy utility `TR_PatternBuilder` which is a very concise and readable way of doing tree pattern matching. There are several examples if you search the code base (eg. [1]).

[1] https://github.com/eclipse/omr/blob/88fd3ecdbbdc0ef7ace859dacc069b2c38e9ef27/compiler/codegen/OMRTreeEvaluator.cpp#L363-L380

Reply:
Thanks, Filip!  I did not know, and I had been thinking that something like that was needed.  Why doesn't its use seem to be particularly widespread?

Reply:
Probably because not many people know about it. I've been encouraging it's use in PR reviews if I spot an opportunity. If it works well for you perhaps you can advocate for it as well? Thanks!

Reply:
@fjeremic Filip, I was playing around with the `TR_PatternBuilder`.  I think in this case using it might be a bit messy because we want to match either `ificmpne` or `iflcmpne`, `iand` or `land`, etc.  So either I would have to build the patterns with `pb.choice` throughout, or build three different patterns:  one pure `int` case, one pure `long` case and one mixed with an `i2l` in the middle.  We might want to think about having some built-in mechanisms for matching either `int` or `long` operations to simplify this kind of scenario.

The other thing I found is that tree pattern matching seems to be currently disabled.  We'll investigate why that is, but Andrew @andrewcraik  suggested I leave it aside.  If you don't have any objection, that's what I'll do for now.  I will definitely look for further opportunities to use the pattern builder once it's enabled.

(As an aside, I really miss XPath!)

Reply:
That sounds reasonable.

Reply:
I think the disable of the matching is conflating the disable of OverflowCHK usage with the pattern matcher - we'll need to separate the two concepts to allow this stuff to be used elsewhere.